pyro/distributions/lkj.py

```diff
@@ -49,10 +47,10 @@ def _call(self, x):
         return x.tanh()
 
     def _inverse(self, y):
-        return torch.log((1 + y) / (1 - y)) / 2
+        return ((1 + y) / (1 - y)).log() / 2
 
     def log_abs_det_jacobian(self, x, y):
-        return (-2) * y.cosh().log()
+        return (1 - y ** 2).log()
```

Comment:
this is a bug in Stan manual!

Reply:
I think that `y**2` is more numerically stable than `x.cosh()`.

About testing, could you suggest some else tests which I missed? Everything looks fine according to the current tests (including batch, scalar, singleton,...)

Reply:
We don't need to take the sum. ComposedTransform does it automatically if we specify a correct `event_dim`.

Reply:
I may be mistaken about this --- but in thinking about it, its hard to see a way to get the calculation right when tanh is a separate transform. It looks to me that the solution is to go back to the way it was setup before, where the transformation from reals->cholesky is a subclass of the transformation from [-1,1]->cholesky, so that we know the event dimension when lob_abs_det_jacobian gets called.  

If you want to write a reals->[-1,1] transformation using tanh (and I'm not sure why you would want to do this -- the usual way to enforce a range transformation is with a modified logit), then I think that should take place separately from this project. 

Anyway, let's try to deal with changes to the tanh part of the code separately from the changes to the cholesky transformation, in a separate PR.  OK?

Reply:
Ok I see your response now. If you let ComposedTransform take care of it, then aren't you getting a less efficient and less numerically stable gradient?

Reply:
If you think something is more stable, then could you please add a test? I can't identify what causes "less numerically stable gradient" here. When I added tests for your code, it didn't pass, so I can't compare. So what I take from your argument is: Stan guys don't do Tanh transformation as in their code?

Reply:
(I have written template with links to pytorch's transforms for you. You can check it to see what x, y play. Here x is in domain, and y is in codomain.)

I don't see what you mean. Could you elaborate on it? Assume we don't separate but combine two transforms into one class, then will you compute `tanh(x)` in log_abs_det_jabobian? If yes, then why we don't use it but computing cosh?

Reply:
Your approach to doing it applies more operations, and does so to more components of the tensor, without any observable benefit.  

While `x**2` is more stable than `cosh(x)`, `1 - tanh(x)**2` is not more stable than `cosh(x)`. 

I'm not saying to combine the two transforms into one class.  I'm saying the transformation from `real->lowercholesky` can subclass the transformation from `[-1,1]->lowercholesky`. (Or it could be a separate transform that then calls `[-1,1]->lowercholesky` internally, that would be fine too.)

Reply:
Because I can't understand, so it is better to see your implementation instead.